org/mozilla/jss/ssl/javax/JSSSocket.java

```diff
@@ -0,0 +1,869 @@
+package org.mozilla.jss.ssl.javax;
+
+import java.io.*;
+import java.net.*;
+import java.nio.channels.*;
+import java.security.*;
+import java.util.*;
+import javax.net.ssl.*;
+
+import org.mozilla.jss.pkcs11.PK11Cert;
+import org.mozilla.jss.pkcs11.PK11PrivKey;
+import org.mozilla.jss.provider.javax.crypto.JSSKeyManager;
+import org.mozilla.jss.provider.javax.crypto.JSSTrustManager;
+
+/**
+ * SSL-enabled socket following the javax.net.ssl.SSLSocket interface.
+ *
+ * Most users will want to use the JSSSocketFactory provided by the Java
+ * Provider interface instead of using this class directly.
+ *
+ * This SSLSocket implementation is a wrapped implementation. In particular, we
+ * need to consume an existing Socket (via the consumeSocket(...) call) which
+ * we actually send data over. When called from a socket factory, this
+ * additional socket will be automatically created for the caller. This
+ * is necessary because SSLSocketFactory includes a mode which wraps an
+ * existing socket.
+ *
+ * All JSSSocket instances have a underlying SocketChannel, of type
+ * JSSSocketChannel. Notably lacking is a javax.net.ssl.SSLSocketChannel type,
+ * so JSSSocketChannel includes no additional SSL-specific options. However,
+ * the core of the SSLEngine wrapping logic exists there.
+ *
+ * In order to interoperate with JSSEngine, many of the adjacent methods have
+ * been included in this class as well.
+ *
+ * This socket can either be a client or a server, depending on how it was
+ * created. For more information, see the javax.net.ssl.SSLSocket
+ * documentation.
+ *
+ * To construct a (useful) new instance, the following calls must be made:
+ *
+ * - new JSSSocket();
+ * - consumeSocket(inst);
+ * - initSSLEngine(...);
+ * - setKeyManagers(...);
+ * - setTrustManagers(...);
+ *
+ * Optionally, setSSLContext(...) could be called to provide the SSLContext
+ * from which the SSLEngine should be constructed. This should be called prior
+ * to initSSLEngine(...) being called.
+ */
+public class JSSSocket extends SSLSocket {
```

Comment:
In the future we probably could change `org.mozilla.jss.ssl.SSLSocket` to extend `JSSSocket`, and eventually merge it into `JSSSocket`.

Reply:
`org.mozilla.jss.ssl.SSLSocket` doesn't respect many of the semantics around the `javax.net.ssl` suite:

 - `KeyManager` for key selection,
 - `TrustManager` for cert validation,
 - Converting between `SSLCipher` and JDK-standard names is easier.

We could fix the latter, but the former would be more invasive changes. Ideally we'd like our `javax.net.ssl.SSLSocket` implementation to behave as similarly to our `javax.net.ssl.SSLEngine` as possible. For people not expecting a `javax.net.ssl.SSLSocket` and who want a generic `Socket` (that happens to do TLS under the covers), we're fine keeping `org.mozilla.jss.ssl.SSLSocket`. I guess that's LDAP JDK and some Dogtag PKI helpers. But for any third-party library, I think `JSSSocket` is fine? I dunno. At any rate, not a problem to solve in this PR :-)